File: hexameter-core/src/main/java/org/codetome/hexameter/core/internal/impl/layoutstrategy/HexagonalGridLayoutStrategy.java
package org.codetome.hexameter.core.internal.impl.layoutstrategy;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
import org.codetome.hexameter.core.api.CubeCoordinate;
import org.codetome.hexameter.core.api.HexagonOrientation;
import org.codetome.hexameter.core.api.HexagonalGrid;
import org.codetome.hexameter.core.api.HexagonalGridBuilder;
import rx.Observable;
import rx.Observable.OnSubscribe;
import rx.Subscriber;

import static java.lang.Math.abs;
import static java.lang.Math.floor;
import static java.lang.Math.max;
import static java.lang.Math.round;

/**
 * This strategy is responsible for generating a {@link HexagonalGrid} which has a hexagonal
 * shape.
 */
@SuppressFBWarnings("SIC_INNER_SHOULD_BE_STATIC_ANON")
public final class HexagonalGridLayoutStrategy extends GridLayoutStrategy {

    @Override
    public Observable<CubeCoordinate> fetchGridCoordinates(final HexagonalGridBuilder builder) {
        Observable<CubeCoordinate> result = Observable.create(new OnSubscribe<CubeCoordinate>() {
            @Override
            public void call(Subscriber<? super CubeCoordinate> subscriber) {
                final double gridSize = builder.getGridHeight();
                int startX = HexagonOrientation.FLAT_TOP.equals(builder.getOrientation()) ? (int) floor(gridSize / 2d) : (int) round(gridSize / 4d);
                final int hexRadius = (int) floor(gridSize / 2d);
                final int minX = startX - hexRadius;
                for (int y = 0; y < gridSize; y++) {
                    final int distanceFromMid = abs(hexRadius - y);
                    for (int x = max(startX, minX); x <= max(startX, minX) + hexRadius + hexRadius - distanceFromMid; x++) {
                        final int gridX = x;
                        final int gridZ = HexagonOrientation.FLAT_TOP.equals(builder.getOrientation()) ? y - (int) floor(gridSize / 4d) : y;
                        subscriber.onNext(CubeCoordinate.fromCoordinates(gridX, gridZ));
                    }
                    startX--;
                }
                subscriber.onCompleted();
            }
        });
        return result;
    }

    @Override
    public boolean checkParameters(final int gridHeight, final int gridWidth) {
        final boolean superResult = super.checkParameters(gridHeight, gridWidth);
        final boolean result = gridHeight == gridWidth && abs(gridHeight % 2) == 1;
        return result && superResult;
    }

}

File: mixite2.core/src/commonMain/kotlin/org/hexworks/mixite2/core/internal/impl/layoutstrategy/HexagonalGridLayoutStrategy.kt
package org.hexworks.mixite2.core.internal.impl.layoutstrategy

import org.hexworks.mixite2.core.api.CubeCoordinate
import org.hexworks.mixite2.core.api.HexagonOrientation
import org.hexworks.mixite2.core.api.HexagonalGridBuilder
import org.hexworks.mixite2.core.api.contract.SatelliteData
import kotlin.math.abs
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.round

class HexagonalGridLayoutStrategy : GridLayoutStrategy() {

    override fun fetchGridCoordinates(builder: HexagonalGridBuilder<out SatelliteData>): Iterable<org.hexworks.mixite2.core.api.CubeCoordinate> {
        val gridSize = builder.getGridHeight()
        val coords = ArrayList<org.hexworks.mixite2.core.api.CubeCoordinate>(1 + (gridSize * gridSize * 6 - 6) / 8) // TODO cell count owned by the builder
        var startX = if (HexagonOrientation.FLAT_TOP.equals(builder.getOrientation())) floor(gridSize / 2.0).toInt() else round(gridSize / 4.0).toInt()
        val hexRadius = floor(gridSize / 2.0).toInt()
        val minX = startX - hexRadius
        var y = 0
        while (y < gridSize) {
            val distanceFromMid = abs(hexRadius - y)
            for (x in max(startX, minX)..max(startX, minX) + hexRadius + hexRadius - distanceFromMid) {
                val gridZ = if (HexagonOrientation.FLAT_TOP.equals(builder.getOrientation())) y - floor(gridSize / 4.0).toInt() else y
                coords.add(org.hexworks.mixite2.core.api.CubeCoordinate.fromCoordinates(x, gridZ))
            }
            startX--
            y++
        }
        return coords
    }

    override fun checkParameters(gridHeight: Int, gridWidth: Int): Boolean {
        val superResult = checkCommonCase(gridHeight, gridWidth)
        val result = gridHeight == gridWidth && abs(gridHeight % 2) == 1
        return result && superResult
    }

    override fun getName(): String {
        return "HEXAGONAL"
    }

}
